### PR TITLE
Optimize the initialization of the cloudcore module global variable

### PR DESCRIPTION
**Pull request by kadisi**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind test
> /kind failing-test
> /kind feature

/kind cleanup
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Optimize the initialization of the cloudcore module global variable
```

### CODE REVIEW

File: cloud/pkg/cloudhub/config/config.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		defer func() {
+			if len(errs) != 0 {
+				for _, e := range errs {
+					klog.Errorf("%v", e)
+				}
+				klog.Error("init cloudhub config error")
+				os.Exit(1)
+			} else {
+				klog.Infof("init cloudhub config successfully，config info %++v", c)
+			}
```
Comment (fisherxu) [inline cloud/pkg/cloudhub/config/config.go]: Remove the defer and move lines behind #115.  And no need else here.

File: cloud/pkg/devicecontroller/config/config.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		defer func() {
+			if len(errs) != 0 {
+				for _, e := range errs {
+					klog.Errorf("%v", e)
+				}
+				klog.Error("init devicecontroller config error")
+				os.Exit(1)
+			} else {
+				klog.Infof("init devicecontroller config successfully, config info %++v", c)
+			}
```
Comment (fisherxu) [inline cloud/pkg/devicecontroller/config/config.go]: ditto

File: cloud/pkg/edgecontroller/config/config.go [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+		defer func() {
+			if len(errs) != 0 {
+				for _, e := range errs {
+					klog.Errorf("%v", e)
+				}
+				klog.Error("init edgecontroller config error")
+				os.Exit(1)
+			} else {
+				klog.Infof("init edgecontroller config successfully, config info %++v", c)
+			}
```
Comment (fisherxu) [inline cloud/pkg/edgecontroller/config/config.go]: ditto

File: cloud/pkg/edgecontroller/config/config.go [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+		}()
+
+		psb, err := config.CONFIG.GetValue("controller.buffer.update-pod-status").ToInt()
+		if err != nil {
+			// Guaranteed forward compatibility @kadisi
+			psb = constants.DefaultUpdatePodStatusBuffer
+			klog.Infof("can not get controller.buffer.update-pod-status key, use default value %v", psb)
```
Comment (fisherxu) [inline cloud/pkg/edgecontroller/config/config.go]: warning?

File: cloud/pkg/cloudhub/config/config.go (hunk 1 above)
Comment (kadisi) [reply cloud/pkg/cloudhub/config/config.go]: ok

File: cloud/pkg/devicecontroller/config/config.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+		}()
+		km, err := config.CONFIG.GetValue("devicecontroller.kube.master").ToString()
+		if err != nil {
+			errs = append(errs,
+				fmt.Errorf("get devicecontroller.kube.master configuration key error %v", err))
```
[the remark this answers is not in the document]
Comment (kadisi) [reply cloud/pkg/devicecontroller/config/config.go]: done

File: cloud/pkg/cloudhub/config/config.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	KeepaliveInterval  int
+	Ca                 []byte
+	Cert               []byte
+	Key                []byte
+	WriteTimeout       int
+	NodeLimit          int
+}
+
+func InitConfigure() {
+	once.Do(func() {
```
Comment (fisherxu) [inline cloud/pkg/cloudhub/config/config.go]: Do we need the once here?

File: cloud/pkg/devicecontroller/config/config.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+	// KubeBurst default is 10
+	KubeBurst int
+	// UpdateDeviceStatusWorkers is the count of goroutines of update device status
+	UpdateDeviceStatusWorkers int
+	// MessageLayer used, context or ssmq, default is context
+	MessageLayer string
+}
+
+func InitConfigure() {
+	once.Do(func() {
```
Comment (fisherxu) [inline cloud/pkg/devicecontroller/config/config.go]: ditto

File: cloud/pkg/edgecontroller/config/config.go [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+	QueryNodeWorkers int
+	// UpdateNodeWorkers is the count of goroutines of update node
+	UpdateNodeWorkers int
+
+	// MessageLayer used, context or ssmq, default is context
+	MessageLayer string
+}
+
+func InitConfigure() {
+	once.Do(func() {
```
Comment (fisherxu) [inline cloud/pkg/edgecontroller/config/config.go]: ditto

File: cloud/pkg/cloudhub/config/config.go (hunk 6 above)
Comment (kadisi) [reply cloud/pkg/cloudhub/config/config.go]: yes, Prevent global variable duplicate assignment.

File: tests/e2e/utils/test_setup.go [hunk 9]
```diff
@@ -35,12 +37,21 @@ func GenerateCerts() error {
...
 	}
 	//Expect(err).Should(BeNil())
 	time.Sleep(5 * time.Second)
+
+	checkcmd := exec.Command("sh", "-c", constants.CheckCloudcore)
+	if err := PrintCombinedOutput(checkcmd); err != nil {
+		catcmd := exec.Command("sh", "-c", constants.CatCloudcoreLog)
+		Infof("===========> Executing: %s\n", strings.Join(catcmd.Args, " "))
+		bytes, _ := catcmd.CombinedOutput()
+		Errorf("find cloudcore process exist, cloudcore log:\n %v", string(bytes))
```
Comment (fisherxu) [inline tests/e2e/utils/test_setup.go]: Failed to run cloudcore, error log:

File: tests/e2e/utils/test_setup.go [hunk 10]
```diff
@@ -52,6 +63,15 @@ func StartEdgeCore() error {
 	}
 	//Expect(err).Should(BeNil())
 	time.Sleep(5 * time.Second)
+
+	checkcmd := exec.Command("sh", "-c", constants.CheckEdgecore)
+	if err := PrintCombinedOutput(checkcmd); err != nil {
+		catcmd := exec.Command("sh", "-c", constants.CatEdgecoreLog)
+		Infof("===========> Executing: %s\n", strings.Join(catcmd.Args, " "))
+		bytes, _ := catcmd.CombinedOutput()
+		Errorf("find edgecore process exist, edgecore log:\n %v", string(bytes))
```
Comment (fisherxu) [inline tests/e2e/utils/test_setup.go]: ditto

File: tests/e2e/utils/test_setup.go [hunk 11]
```diff
@@ -84,5 +104,14 @@ func StartEdgeSite() error {
 	}
 	//Expect(err).Should(BeNil())
 	time.Sleep(5 * time.Second)
+
+	checkcmd := exec.Command("sh", "-c", constants.CheckEdgesite)
+	if err := PrintCombinedOutput(checkcmd); err != nil {
+		catcmd := exec.Command("sh", "-c", constants.CatEdgeSiteLog)
+		Infof("===========> Executing: %s\n", strings.Join(catcmd.Args, " "))
+		bytes, _ := catcmd.CombinedOutput()
+		Errorf("find edgesite process exist, edgesite log:\n %v", string(bytes))
```
Comment (fisherxu) [inline tests/e2e/utils/test_setup.go]: ditto

File: cloud/pkg/cloudhub/config/config.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+			Ca:                 ca,
+			Cert:               cert,
+			Key:                key,
+		}
+		if !c.ProtocolWebsocket && !c.ProtocolQuic {
+			c.ProtocolWebsocket = true
+		}
+	})
+}
+func Get() Configure {
```
Comment (fisherxu) [inline cloud/pkg/cloudhub/config/config.go]: return pointer

File: tests/e2e/utils/test_setup.go (hunk 9 above)
Comment (kadisi) [reply tests/e2e/utils/test_setup.go]: done

File: tests/e2e/utils/test_setup.go (hunk 10 above)
Comment (kadisi) [reply tests/e2e/utils/test_setup.go]: done

File: tests/e2e/utils/test_setup.go (hunk 11 above)
Comment (kadisi) [reply tests/e2e/utils/test_setup.go]: done

File: cloud/pkg/cloudhub/config/config.go (hunk 12 above)
Comment (kadisi) [reply cloud/pkg/cloudhub/config/config.go]: done

**Comment by fisherxu** [comment]
/lgtm
Thanks for the effort.

File: tests/e2e/constants/constants.go [hunk 13]
```diff
@@ -1,9 +1,17 @@
...
 
 const (
-	RunController = "cd ${GOPATH}/src/github.com/kubeedge/kubeedge/cloud/; sudo nohup ./cloudcore > cloudcore.log 2>&1 &"
-	RunEdgecore   = "cd ${GOPATH}/src/github.com/kubeedge/kubeedge/edge/; sudo nohup ./edgecore > edgecore.log 2>&1 &"
-	RunEdgeSite   = "cd ${GOPATH}/src/github.com/kubeedge/kubeedge/edgesite/; sudo nohup ./edgesite > edgesite.log 2>&1 &"
+	RunCloudcore = "cd ${GOPATH}/src/github.com/kubeedge/kubeedge/cloud/; sudo nohup ./cloudcore > cloudcore.log 2>&1 &"
+	RunEdgecore  = "cd ${GOPATH}/src/github.com/kubeedge/kubeedge/edge/; sudo nohup ./edgecore > edgecore.log 2>&1 &"
+	RunEdgeSite  = "cd ${GOPATH}/src/github.com/kubeedge/kubeedge/edgesite/; sudo nohup ./edgesite > edgesite.log 2>&1 &"
+
+	CheckCloudcore = "sudo pkill -0 cloudcore"
```
Comment (kevin-wangzefeng) [inline tests/e2e/constants/constants.go]: why use pkill inside a check? you can record the pid when starting components and checkif the pid is still alive

Comment (kadisi) [reply tests/e2e/constants/constants.go]: changed to `pgrep`

**Pull request merged**